katharsis-core/src/main/java/io/katharsis/queryspec/InMemoryEvaluator.java

```diff
@@ -97,11 +97,15 @@ private static boolean matchesPrimitiveOperator(Object object, FilterSpec filter
 		Object value = PropertyUtils.getProperty(object, filterSpec.getAttributePath());
 		FilterOperator operator = filterSpec.getOperator();
 		Object filterValue = filterSpec.getValue();
-		if (value instanceof Collection) {
-			return matchesAny((Collection<?>) value, operator, filterValue);
-		}
-		else {
-			return operator.matches(value, filterValue);
+		if (filterValue != null ||  (filterValue == null && (operator.equals(FilterOperator.EQ) || operator.equals(FilterOperator.NEQ)))) {
```

Comment:
Please refactor to be more like: 

```
boolean result = false;
if(...) {
   result = matchesAny(...)
}
return result;
```
This gives one exit from the function and you can ditch the totally redunant `else's here. 

your code `if(...) return matchesAny(...)` exits the function. Making the subsequent `else` pointless. You can make it just `result =  operator.matches(...);`

finally `else return false` can be replaced with `return result`

Finally: `result = value instanceof Collection ? ((Collection)value).stream().filter(v -> operator.matches(v,filterValue).findAny().orElse(false)` you could eliminate all of the branching